-f-template-haskell causes build failure

It's an upstream report of a downstream failure reported by Toralf Förster  in https://bugs.gentoo.org/800248

When `linear` is built with `-f-template-haskell` it fails to use hidden `template-haskell` modules:

```
Using Parsec parser
Configuring linear-1.21.5...
Flags chosen: herbie=False, template-haskell=False
Dependency adjunctions ==4.*: using adjunctions-4.4
Dependency base >=4.5 && <5: using base-4.14.2.0
Dependency base-orphans >=0.8.3 && <1: using base-orphans-0.8.4
Dependency binary >=0.5 && <0.9: using binary-0.8.8.0
Dependency bytes >=0.15 && <1: using bytes-0.17.1
Dependency cereal >=0.4.1.1 && <0.6: using cereal-0.5.8.1
Dependency containers >=0.4 && <0.7: using containers-0.6.4.1
Dependency deepseq >=1.1 && <1.5: using deepseq-1.4.4.0
Dependency distributive >=0.2.2 && <1: using distributive-0.6.2.1
Dependency ghc-prim -any: using ghc-prim-0.6.1
Dependency hashable >=1.1 && <1.4: using hashable-1.3.2.0
Dependency indexed-traversable >=0.1.1 && <0.2: using
indexed-traversable-0.1.1
Dependency lens >=4.15.2 && <6: using lens-5.0.1
Dependency random >=1.0 && <1.3: using random-1.1
Dependency reflection >=1.3.2 && <3: using reflection-2.1.6
Dependency semigroupoids >=5.2.1 && <6: using semigroupoids-5.3.5
Dependency semigroups >=0.9 && <1: using semigroups-0.18.5
Dependency tagged >=0.4.4 && <1: using tagged-0.8.6.1
Dependency transformers >=0.2 && <0.6: using transformers-0.5.6.2
Dependency transformers-compat >=0.4 && <1: using transformers-compat-0.6.6
Dependency unordered-containers >=0.2.3 && <0.3: using
unordered-containers-0.2.14.0
Dependency vector >=0.10 && <0.13: using vector-0.12.3.0
Dependency void >=0.6 && <1: using void-0.7.3
Dependency base -any: using base-4.14.2.0
Dependency simple-reflect >=0.3.1: using simple-reflect-0.3.3
Dependency HUnit >=1.2.5: using HUnit-1.6.2.0
Dependency base -any: using base-4.14.2.0
Dependency binary -any: using binary-0.8.8.0
Dependency bytestring -any: using bytestring-0.10.12.0
Dependency deepseq -any: using deepseq-1.4.4.0
Dependency linear -any: using linear-1.21.5
Dependency reflection -any: using reflection-2.1.6
Dependency test-framework >=0.8: using test-framework-0.8.2.0
Dependency test-framework-hunit >=0.3: using test-framework-hunit-0.3.0.2
Dependency vector -any: using vector-0.12.3.0
Source component graph:
    component test:doctests
    component lib
    component test:UnitTests dependency lib
Configured component graph:
    component linear-1.21.5-HXIiWNo7x1r4DMqC8Jk03V-doctests
        include base-4.14.2.0
        include simple-reflect-0.3.3-C3REZ52VTBC6wXIpKK13lR
    component linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb
        include adjunctions-4.4-LHlfQPb2PBjBta5MG3NNky
        include base-4.14.2.0
        include base-orphans-0.8.4-B4zsXTdczgN9mmNUwgnmE3
        include binary-0.8.8.0
        include bytes-0.17.1-EiIngf45OZQClWAnATl37M
        include cereal-0.5.8.1-QSBz4OdKmpCBspe656QT0
        include containers-0.6.4.1
        include deepseq-1.4.4.0
        include distributive-0.6.2.1-Ybn7eEKKiB9UWWwlqsXyV
        include ghc-prim-0.6.1
        include hashable-1.3.2.0-IfK3qW7NPQrKxQR5xhhorE
        include indexed-traversable-0.1.1-1Z2GfU1EWlL3ZzMnGNHsA4
        include lens-5.0.1-CH98nPbbU2a3M5qzgghfwy
        include random-1.1-EAzDG87Oakm6nTs5elGBui
        include reflection-2.1.6-CF5M5ZmtpznJcGqaAQnFg7
        include semigroupoids-5.3.5-8E06UZB7MK315SLpS1Mh4T
        include semigroups-0.18.5-75PHB5lYhV2Y86DizAuAl
        include tagged-0.8.6.1-F25vmSN4H5fEaMJYsLgYMV
        include transformers-0.5.6.2
        include transformers-compat-0.6.6-CjrqNEH0LxRHB2LMdBXE5Y
        include unordered-containers-0.2.14.0-5nzYDtC43AV9aNsnSwXPjZ
        include vector-0.12.3.0-LNrFMvx9fFG9u3uIvsIv7B
        include void-0.7.3-72JC1ZJs2iULdltk7dpkuA
    component linear-1.21.5-C5SZEEVraEe3clEWCpBIeq-UnitTests
        include HUnit-1.6.2.0-9r11msDZDxYIaPvKAteuQg
        include base-4.14.2.0
        include binary-0.8.8.0
        include bytestring-0.10.12.0
        include deepseq-1.4.4.0
        include linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb
        include reflection-2.1.6-CF5M5ZmtpznJcGqaAQnFg7
        include test-framework-0.8.2.0-JTSLAgq45PV73f1irT17Gp
        include test-framework-hunit-0.3.0.2-6JnfxavG0qy7kC4uAhgUO2
        include vector-0.12.3.0-LNrFMvx9fFG9u3uIvsIv7B
Linked component graph:
    unit linear-1.21.5-HXIiWNo7x1r4DMqC8Jk03V-doctests
        include base-4.14.2.0
        include simple-reflect-0.3.3-C3REZ52VTBC6wXIpKK13lR
    unit linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb
        include adjunctions-4.4-LHlfQPb2PBjBta5MG3NNky
        include base-4.14.2.0
        include base-orphans-0.8.4-B4zsXTdczgN9mmNUwgnmE3
        include binary-0.8.8.0
        include bytes-0.17.1-EiIngf45OZQClWAnATl37M
        include cereal-0.5.8.1-QSBz4OdKmpCBspe656QT0
        include containers-0.6.4.1
        include deepseq-1.4.4.0
        include distributive-0.6.2.1-Ybn7eEKKiB9UWWwlqsXyV
        include ghc-prim-0.6.1
        include hashable-1.3.2.0-IfK3qW7NPQrKxQR5xhhorE
        include indexed-traversable-0.1.1-1Z2GfU1EWlL3ZzMnGNHsA4
        include lens-5.0.1-CH98nPbbU2a3M5qzgghfwy
        include random-1.1-EAzDG87Oakm6nTs5elGBui
        include reflection-2.1.6-CF5M5ZmtpznJcGqaAQnFg7
        include semigroupoids-5.3.5-8E06UZB7MK315SLpS1Mh4T
        include semigroups-0.18.5-75PHB5lYhV2Y86DizAuAl
        include tagged-0.8.6.1-F25vmSN4H5fEaMJYsLgYMV
        include transformers-0.5.6.2
        include transformers-compat-0.6.6-CjrqNEH0LxRHB2LMdBXE5Y
        include unordered-containers-0.2.14.0-5nzYDtC43AV9aNsnSwXPjZ
        include vector-0.12.3.0-LNrFMvx9fFG9u3uIvsIv7B
        include void-0.7.3-72JC1ZJs2iULdltk7dpkuA
        Linear=linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb:Linear,Linear.Affine=linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb:Linear.Affine,Linear.Algebra=linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb:Linear.Algebra,Linear.Binary=linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb:Linear.Binary,Linear.Conjugate=linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb:Linear.Conjugate,Linear.Covector=linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb:Linear.Covector,Linear.Epsilon=linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb:Linear.Epsilon,Linear.Instances=linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb:Linear.Instances,Linear.Matrix=linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb:Linear.Matrix,Linear.Metric=linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb:Linear.Metric,Linear.Plucker=linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb:Linear.Plucker,Linear.Plucker.Coincides=linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb:Linear.Plucker.Coincides,Linear.Projection=linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb:Linear.Projection,Linear.Quaternion=linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb:Linear.Quaternion,Linear.Trace=linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb:Linear.Trace,Linear.V=linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb:Linear.V,Linear.V0=linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb:Linear.V0,Linear.V1=linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb:Linear.V1,Linear.V2=linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb:Linear.V2,Linear.V3=linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb:Linear.V3,Linear.V4=linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb:Linear.V4,Linear.Vector=linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb:Linear.Vector
    unit linear-1.21.5-C5SZEEVraEe3clEWCpBIeq-UnitTests
        include HUnit-1.6.2.0-9r11msDZDxYIaPvKAteuQg
        include base-4.14.2.0
        include binary-0.8.8.0
        include bytestring-0.10.12.0
        include deepseq-1.4.4.0
        include linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb
        include reflection-2.1.6-CF5M5ZmtpznJcGqaAQnFg7
        include test-framework-0.8.2.0-JTSLAgq45PV73f1irT17Gp
        include test-framework-hunit-0.3.0.2-6JnfxavG0qy7kC4uAhgUO2
        include vector-0.12.3.0-LNrFMvx9fFG9u3uIvsIv7B
Ready component graph:
    definite linear-1.21.5-HXIiWNo7x1r4DMqC8Jk03V-doctests
        depends base-4.14.2.0
        depends simple-reflect-0.3.3-C3REZ52VTBC6wXIpKK13lR
    definite linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb
        depends adjunctions-4.4-LHlfQPb2PBjBta5MG3NNky
        depends base-4.14.2.0
        depends base-orphans-0.8.4-B4zsXTdczgN9mmNUwgnmE3
        depends binary-0.8.8.0
        depends bytes-0.17.1-EiIngf45OZQClWAnATl37M
        depends cereal-0.5.8.1-QSBz4OdKmpCBspe656QT0
        depends containers-0.6.4.1
        depends deepseq-1.4.4.0
        depends distributive-0.6.2.1-Ybn7eEKKiB9UWWwlqsXyV
        depends ghc-prim-0.6.1
        depends hashable-1.3.2.0-IfK3qW7NPQrKxQR5xhhorE
        depends indexed-traversable-0.1.1-1Z2GfU1EWlL3ZzMnGNHsA4
        depends lens-5.0.1-CH98nPbbU2a3M5qzgghfwy
        depends random-1.1-EAzDG87Oakm6nTs5elGBui
        depends reflection-2.1.6-CF5M5ZmtpznJcGqaAQnFg7
        depends semigroupoids-5.3.5-8E06UZB7MK315SLpS1Mh4T
        depends semigroups-0.18.5-75PHB5lYhV2Y86DizAuAl
        depends tagged-0.8.6.1-F25vmSN4H5fEaMJYsLgYMV
        depends transformers-0.5.6.2
        depends transformers-compat-0.6.6-CjrqNEH0LxRHB2LMdBXE5Y
        depends unordered-containers-0.2.14.0-5nzYDtC43AV9aNsnSwXPjZ
        depends vector-0.12.3.0-LNrFMvx9fFG9u3uIvsIv7B
        depends void-0.7.3-72JC1ZJs2iULdltk7dpkuA
    definite linear-1.21.5-C5SZEEVraEe3clEWCpBIeq-UnitTests
        depends HUnit-1.6.2.0-9r11msDZDxYIaPvKAteuQg
        depends base-4.14.2.0
        depends binary-0.8.8.0
        depends bytestring-0.10.12.0
        depends deepseq-1.4.4.0
        depends linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb
        depends reflection-2.1.6-CF5M5ZmtpznJcGqaAQnFg7
        depends test-framework-0.8.2.0-JTSLAgq45PV73f1irT17Gp
        depends test-framework-hunit-0.3.0.2-6JnfxavG0qy7kC4uAhgUO2
        depends vector-0.12.3.0-LNrFMvx9fFG9u3uIvsIv7B
Using Cabal-3.2.1.0 compiled by ghc-8.10
Using compiler: ghc-8.10.5
Using install prefix: /usr/local
Executables installed in: /usr/local/bin
Libraries installed in:
/usr/local/lib/x86_64-linux-ghc-8.10.5/linear-1.21.5-8A0VCFTcc7l3FcdXgHRvMb
Dynamic Libraries installed in: /usr/local/lib/x86_64-linux-ghc-8.10.5
Private executables installed in:
/usr/local/libexec/x86_64-linux-ghc-8.10.5/linear-1.21.5
Data files installed in:
/usr/local/share/x86_64-linux-ghc-8.10.5/linear-1.21.5
Documentation installed in:
/usr/local/share/doc/x86_64-linux-ghc-8.10.5/linear-1.21.5
Configuration files installed in: /usr/local/etc
Using alex version 3.2.6 found on system at: /usr/bin/alex
Using ar found on system at: /usr/bin/x86_64-pc-linux-gnu-ar
Using c2hs version 0.28.8 found on system at: /usr/bin/c2hs
Using cpphs version 1.20.9 found on system at: /usr/bin/cpphs
Using doctest version 0.18.1 found on system at: /usr/bin/doctest
Using gcc version 12.0.0 found on system at:
/usr/lib/ccache/bin/x86_64-pc-linux-gnu-gcc
Using ghc version 8.10.5 found on system at: /usr/bin/ghc
Using ghc-pkg version 8.10.5 found on system at: /usr/bin/ghc-pkg
No ghcjs found
No ghcjs-pkg found
No greencard found
Using haddock version 2.24.0 found on system at: /usr/bin/haddock
Using happy version 1.20.0 found on system at: /usr/bin/happy
Using haskell-suite found on system at: haskell-suite-dummy-location
Using haskell-suite-pkg found on system at: haskell-suite-pkg-dummy-location
No hmake found
Using hpc version 0.68 found on system at: /usr/bin/hpc
Using hsc2hs version 0.68.7 found on system at: /usr/bin/hsc2hs
Using hscolour version 1.24 found on system at: /usr/bin/HsColour
No jhc found
Using ld found on system at: /usr/bin/x86_64-pc-linux-gnu-ld
Using pkg-config version 1.6.3 found on system at: /usr/bin/pkg-config
Using runghc version 8.10.5 found on system at: /usr/bin/runghc
No strip found
Using tar found on system at: /bin/tar
No uhc found
Preprocessing test suite 'doctests' for linear-1.21.5..
Building test suite 'doctests' for linear-1.21.5..
Preprocessing library for linear-1.21.5..
Building library for linear-1.21.5..
[ 8 of 22] Compiling Linear.V1        ( src/Linear/V1.hs, dist/build/Linear/V1.o, dist/build/Linear/V1.dyn_o )

src/Linear/V1.hs:92:1: error:
    Could not load module ‘Language.Haskell.TH.Syntax’
    It is a member of the hidden package ‘template-haskell-2.16.0.0’.
    Perhaps you need to add ‘template-haskell’ to the build-depends in your .cabal file.
    It is a member of the hidden package ‘ghc-lib-parser-8.10.2.20200916’.
    Perhaps you need to add ‘ghc-lib-parser’ to the build-depends in your .cabal file.
    Use -v (or `:set -v` in ghci) to see a list of the files searched for.
   |
92 | import Language.Haskell.TH.Syntax (Lift)
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[12 of 22] Compiling Linear.V0        ( src/Linear/V0.hs, dist/build/Linear/V0.o, dist/build/Linear/V0.dyn_o )

src/Linear/V0.hs:88:1: error:
    Could not load module ‘Language.Haskell.TH.Syntax’
    It is a member of the hidden package ‘template-haskell-2.16.0.0’.
    Perhaps you need to add ‘template-haskell’ to the build-depends in your .cabal file.
    It is a member of the hidden package ‘ghc-lib-parser-8.10.2.20200916’.
    Perhaps you need to add ‘ghc-lib-parser’ to the build-depends in your .cabal file.
    Use -v (or `:set -v` in ghci) to see a list of the files searched for.
   |
88 | import Language.Haskell.TH.Syntax (Lift)
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

--- a/.github/workflows/haskell-ci.yml
+++ b/.github/workflows/haskell-ci.yml
@@ -8,9 +8,9 @@
 #
 # For more information, see https://github.com/haskell-CI/haskell-ci
 #
-# version: 0.12
+# version: 0.12.1
 #
-# REGENDATA ("0.12",["github","cabal.project"])
+# REGENDATA ("0.12.1",["github","cabal.project"])
 #
 name: Haskell-CI
 on:
@@ -140,7 +140,7 @@ jobs:
       - name: cache (tools)
         uses: actions/cache@v2
         with:
-          key: ${{ runner.os }}-${{ matrix.compiler }}-tools-44f6870a
+          key: ${{ runner.os }}-${{ matrix.compiler }}-tools-33871eb4
           path: ~/.haskell-ci-tools
       - name: install cabal-plan
         run: |
@@ -162,7 +162,7 @@ jobs:
           cabal-docspec --version
       - name: install hlint
         run: |
-          if [ $((HCNUMVER >= 81000 && HCNUMVER < 90000)) -ne 0 ] ; then HLINTVER=$(cd /tmp && (${CABAL} v2-install -v $ARG_COMPILER --dry-run hlint  --constraint='hlint >=3.2 && <3.3' |  perl -ne 'if (/\bhlint-(\d+(\.\d+)*)\b/) { print "$1"; last; }')); echo "HLint version $HLINTVER" ; fi
+          if [ $((HCNUMVER >= 81000 && HCNUMVER < 90000)) -ne 0 ] ; then HLINTVER=$(cd /tmp && (${CABAL} v2-install -v $ARG_COMPILER --dry-run hlint  --constraint='hlint >=3.3 && <3.4' |  perl -ne 'if (/\bhlint-(\d+(\.\d+)*)\b/) { print "$1"; last; }')); echo "HLint version $HLINTVER" ; fi
           if [ $((HCNUMVER >= 81000 && HCNUMVER < 90000)) -ne 0 ] ; then if [ ! -e $HOME/.haskell-ci-tools/hlint-$HLINTVER/hlint ]; then echo "Downloading HLint version $HLINTVER"; mkdir -p $HOME/.haskell-ci-tools; curl --write-out 'Status Code: %{http_code} Redirects: %{num_redirects} Total time: %{time_total} Total Dsize: %{size_download}\n' --silent --location --output $HOME/.haskell-ci-tools/hlint-$HLINTVER.tar.gz "https://github.com/ndmitchell/hlint/releases/download/v$HLINTVER/hlint-$HLINTVER-x86_64-linux.tar.gz"; tar -xzv -f $HOME/.haskell-ci-tools/hlint-$HLINTVER.tar.gz -C $HOME/.haskell-ci-tools; fi ; fi
           if [ $((HCNUMVER >= 81000 && HCNUMVER < 90000)) -ne 0 ] ; then mkdir -p $CABAL_DIR/bin && ln -sf "$HOME/.haskell-ci-tools/hlint-$HLINTVER/hlint" $CABAL_DIR/bin/hlint ; fi
           if [ $((HCNUMVER >= 81000 && HCNUMVER < 90000)) -ne 0 ] ; then hlint --version ; fi
@@ -231,3 +231,9 @@ jobs:
       - name: haddock
         run: |
           $CABAL v2-haddock $ARG_COMPILER --with-haddock $HADDOCK $ARG_TESTS $ARG_BENCH all
+      - name: prepare for constraint sets
+        run: |
+          rm -f cabal.project.local
+      - name: constraint set no-template-haskell
+        run: |
+          $CABAL v2-build $ARG_COMPILER --disable-tests --disable-benchmarks --constraint='linear -template-haskell' all

--- a/cabal.haskell-ci
+++ b/cabal.haskell-ci
@@ -1,3 +1,4 @@
+distribution:           bionic
 no-tests-no-benchmarks: False
 unconstrained:          False
 hlint:                  True
@@ -5,3 +6,6 @@ hlint-job:              8.10.4
 irc-channels:           irc.freenode.org#haskell-lens
 irc-if-in-origin-repo:  True
 docspec:                True
+
+constraint-set no-template-haskell
+  constraints: linear -template-haskell

--- a/src/Linear/Plucker.hs
+++ b/src/Linear/Plucker.hs
@@ -95,7 +95,7 @@ import GHC.Generics (Generic)
 #if __GLASGOW_HASKELL__ >= 706
 import GHC.Generics (Generic1)
 #endif
-#if __GLASGOW_HASKELL__ >= 800
+#if __GLASGOW_HASKELL__ >= 800 && defined(MIN_VERSION_template_haskell)
 import Language.Haskell.TH.Syntax (Lift)
 #endif
 import Linear.Epsilon
@@ -117,7 +117,7 @@ data Plucker a = Plucker !a !a !a !a !a !a deriving (Eq,Ord,Show,Read
 #if __GLASGOW_HASKELL__ >= 706
                                                     ,Generic1
 #endif
-#if __GLASGOW_HASKELL__ >= 800
+#if __GLASGOW_HASKELL__ >= 800 && defined(MIN_VERSION_template_haskell)
                                                     ,Lift
 #endif
                                                     )

--- a/src/Linear/Quaternion.hs
+++ b/src/Linear/Quaternion.hs
@@ -103,7 +103,7 @@ import GHC.Generics (Generic)
 #if __GLASGOW_HASKELL__ >= 706
 import GHC.Generics (Generic1)
 #endif
-#if __GLASGOW_HASKELL__ >= 800
+#if __GLASGOW_HASKELL__ >= 800 && defined(MIN_VERSION_template_haskell)
 import Language.Haskell.TH.Syntax (Lift)
 #endif
 import Linear.Epsilon
@@ -128,7 +128,7 @@ data Quaternion a = Quaternion !a {-# UNPACK #-}!(V3 a)
 #if defined(__GLASGOW_HASKELL__) && __GLASGOW_HASKELL__ >= 706
                              ,Generic1
 #endif
-#if __GLASGOW_HASKELL__ >= 800
+#if __GLASGOW_HASKELL__ >= 800 && defined(MIN_VERSION_template_haskell)
                              ,Lift
 #endif
                              )

--- a/src/Linear/V0.hs
+++ b/src/Linear/V0.hs
@@ -84,7 +84,7 @@ import GHC.Generics (Generic)
 #if __GLASGOW_HASKELL__ >= 706
 import GHC.Generics (Generic1)
 #endif
-#if __GLASGOW_HASKELL__ >= 800
+#if __GLASGOW_HASKELL__ >= 800 && defined(MIN_VERSION_template_haskell)
 import Language.Haskell.TH.Syntax (Lift)
 #endif
 import qualified Data.Vector.Generic.Mutable as M
@@ -120,7 +120,7 @@ data V0 a = V0 deriving (Eq,Ord,Show,Read,Ix,Enum,Data,Typeable
 #if defined(__GLASGOW_HASKELL__) && __GLASGOW_HASKELL__ >= 706
                         ,Generic1
 #endif
-#if __GLASGOW_HASKELL__ >= 800
+#if __GLASGOW_HASKELL__ >= 800 && defined(MIN_VERSION_template_haskell)
                         ,Lift
 #endif
                         )

--- a/src/Linear/V1.hs
+++ b/src/Linear/V1.hs
@@ -88,7 +88,7 @@ import GHC.Generics (Generic)
 #if __GLASGOW_HASKELL__ >= 706
 import GHC.Generics (Generic1)
 #endif
-#if __GLASGOW_HASKELL__ >= 800
+#if __GLASGOW_HASKELL__ >= 800 && defined(MIN_VERSION_template_haskell)
 import Language.Haskell.TH.Syntax (Lift)
 #endif
 import Linear.Metric
@@ -135,7 +135,7 @@ newtype V1 a = V1 a
 #if defined(__GLASGOW_HASKELL__) && __GLASGOW_HASKELL__ >= 706
            ,Generic1
 #endif
-#if __GLASGOW_HASKELL__ >= 800
+#if __GLASGOW_HASKELL__ >= 800 && defined(MIN_VERSION_template_haskell)
            ,Lift
 #endif
            )

--- a/src/Linear/V2.hs
+++ b/src/Linear/V2.hs
@@ -92,7 +92,7 @@ import GHC.Generics (Generic)
 #if __GLASGOW_HASKELL__ >= 706
 import GHC.Generics (Generic1)
 #endif
-#if __GLASGOW_HASKELL__ >= 800
+#if __GLASGOW_HASKELL__ >= 800 && defined(MIN_VERSION_template_haskell)
 import Language.Haskell.TH.Syntax (Lift)
 #endif
 import qualified Data.Vector.Generic.Mutable as M
@@ -136,7 +136,7 @@ data V2 a = V2 !a !a deriving
 #if defined(__GLASGOW_HASKELL__) && __GLASGOW_HASKELL__ >= 706
   ,Generic1
 #endif
-#if __GLASGOW_HASKELL__ >= 800
+#if __GLASGOW_HASKELL__ >= 800 && defined(MIN_VERSION_template_haskell)
   ,Lift
 #endif
   )

--- a/src/Linear/V3.hs
+++ b/src/Linear/V3.hs
@@ -93,7 +93,7 @@ import GHC.Generics (Generic)
 #if __GLASGOW_HASKELL__ >= 706
 import GHC.Generics (Generic1)
 #endif
-#if __GLASGOW_HASKELL__ >= 800
+#if __GLASGOW_HASKELL__ >= 800 && defined(MIN_VERSION_template_haskell)
 import Language.Haskell.TH.Syntax (Lift)
 #endif
 import Linear.Epsilon
@@ -116,7 +116,7 @@ data V3 a = V3 !a !a !a deriving (Eq,Ord,Show,Read,Data,Typeable
 #if __GLASGOW_HASKELL__ >= 706
                                  ,Generic1
 #endif
-#if __GLASGOW_HASKELL__ >= 800
+#if __GLASGOW_HASKELL__ >= 800 && defined(MIN_VERSION_template_haskell)
                                  ,Lift
 #endif
                                  )

--- a/src/Linear/V4.hs
+++ b/src/Linear/V4.hs
@@ -99,7 +99,7 @@ import GHC.Generics (Generic)
 #if __GLASGOW_HASKELL__ >= 706
 import GHC.Generics (Generic1)
 #endif
-#if __GLASGOW_HASKELL__ >= 800
+#if __GLASGOW_HASKELL__ >= 800 && defined(MIN_VERSION_template_haskell)
 import Language.Haskell.TH.Syntax (Lift)
 #endif
 import Linear.Epsilon
@@ -123,7 +123,7 @@ data V4 a = V4 !a !a !a !a deriving (Eq,Ord,Show,Read,Data,Typeable
 #if __GLASGOW_HASKELL__ >= 706
                                     ,Generic1
 #endif
-#if __GLASGOW_HASKELL__ >= 800
+#if __GLASGOW_HASKELL__ >= 800 && defined(MIN_VERSION_template_haskell)
                                     ,Lift
 #endif
                                     )

